src/vnacal_parameter.3
.\"
.\" Vector Network Analyzer Library
.\" Copyright © 2020-2026 D Scott Guthridge <scott_guthridge@rompromity.net>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published
.\" by the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.TH VNACAL_PARAMETER 3 "2022-11-25" GNU
.nh
.SH NAME
vnacal_make_scalar_parameter, vnacal_make_vector_parameter, vnacal_make_calkit_parameter, vnacal_make_calkit_parameter_matrix, vnacal_make_data_parameter, vnacal_make_data_parameter_matrix, vnacal_load_data_parameter_matrix, vnacal_make_unknown_parameter, vnacal_make_correlated_parameter, vnacal_get_parameter_value, vnacal_eval_parameter, vnacal_eval_parameter_matrix, vnacal_delete_parameter, vnacal_delete_parameter_matrix \- calibration parameter construction and evaluation

.\"
.SH SYNOPSIS
.B #include <vnacal.h>
.\"
.PP
.TS
tab(;);
ll.
\fB\s-2VNACAL_MATCH\s+2\fP; (also \fB\s-2VNACAL_ZERO\s+2\fP)
\fB\s-2VNACAL_OPEN\s+2\fP; (also \fB\s-2VNACAL_ONE\s+2\fP)
\fB\s-2VNACAL_SHORT\s+2\fP;
.TE
.\"
.PP
.BI "int vnacal_make_scalar_parameter(vnacal_t *" vcp ,
.if n .RS +4n
.BI "double complex " coefficient );
.if n .RS -4n
.\"
.PP
.BI "int vnacal_make_vector_parameter(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const double *" frequency_vector ,
.if !n .RS +4n
.BI "int " frequencies ,
.if n .br
.BI "const double complex *" coefficient_vector );
.RS -4n
.\"
.PP
.BI "int vnacal_make_calkit_parameter(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const vnacal_calkit_data_t *" vcdp );
.if n .RS -4n
.\"
.PP
.BI "int vnacal_make_calkit_parameter_matrix(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const vnacal_calkit_data_t *" vcdp ,
.if !n .RS +4n
.BI "int *" parameter_matrix ,
.if n .br
.BI "size_t " parameter_matrix_size );
.RS -4
.\"
.PP
.BI "int vnacal_make_data_parameter(vnacal_t *" vcp ", const vnadata_t *" vdp );
.\"
.PP
.BI "int vnacal_make_data_parameter_matrix(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const vnadata_t *" vdp ",
.if !n .RS +4n
.BI "int *" parameter_matrix ,
.if n .br
.BI "size_t " parameter_matrix_size );
.RS -4n
.\"
.PP
.BI "int vnacal_load_data_parameter_matrix(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const *" filename ",
.if !n .RS +4n
.BI "int *" parameter_matrix ,
.if n .br
.BI "size_t " parameter_matrix_size );
.RS -4n
.\"
.PP
.BI "int vnacal_make_unknown_parameter(vnacal_t *" vcp ", int " initial_guess );
.PP
.BI "int vnacal_make_correlated_parameter(vnacal_t *" vcp ", int " other ,
.RS +4n
.BI "const double *" sigma_frequency_vector ", int " sigma_frequencies ,
.if n .br
.BI "const double *" sigma_vector );
.RS -4n
.\"
.PP
.BI "double complex vnacal_get_parameter_value(vnacal_t *" vcp ,
.if n .RS +4n
.BI "int " parameter ", double " frequency );
.if n .RS -4n
.\"
.PP
.BI "double complex vnacal_eval_parameter(vnacal_t *" vcp ,
.if n .RS +4n
.BI "int " parameter ", double " frequency ",
.if !n .RS +4n
.BI "double complex " z0 );
.RS -4n
.\"
.PP
.BI "int vnacal_eval_parameter_matrix(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const int *" parameter_matrix ", int " rows ", int " columns ,
.ie n .br
.el .RS +4n
.BI "double " frequency ", const double complex *" z0_vector ,
.if n .br
.BI "double complex *" result_matrix );
.RS -4n
.\"
.PP
.BI "int vnacal_parameter_matrix_to_data(vnacal_t *" vcp ,
.if n .RS +4n
.BI "const int *" parameter_matrix ,
.if !n .RS +4n
.BI "int " rows ", int " columns ,
.if n .br
.BI "vnadata_t *" vdp );
.RS -4n
.\"
.PP
.BI "int vnacal_delete_parameter(vnacal_t *" vcp ", int " parameter );
.\"
.PP
.BI "int vnacal_delete_parameter_matrix(vnacal_t *" vcp ,
.if n .RS +4n
.BI "int *" parameter_matrix ,
.if !n .RS +4n
.BI "int " rows ", int " columns );
.RS -4n
.\"
.PP
Link with \fI-lvna\fP \fI-lyaml\fP \fI-lm\fP.
.sp
.\"
.SH DESCRIPTION
.PP
These functions construct calibration standard parameter objects as
required for the \fBvnacal_new_add_\fP*() functions.
Instead of taking a matrix of complex numbers that specify the
S-parameters of a standard directly, the \fBvnacal_new_add_\fP*()
functions take a matrix of type \fBint\fP, filled with
either the predefined constants: \fB\s-2VNACAL_MATCH\s+2\fP,
\fB\s-2VNACAL_OPEN\s+2\fP, \fB\s-2VNACAL_SHORT\s+2\fP,
\fB\s-2VNACAL_ZERO\s+2\fP, \fB\s-2VNACAL_ONE\s+2\fP, or integer handles
returned from the \fBvnacal_make_\fP*\fP_parameter\fP() functions
described here.
Each integer value represents a persistent parameter object with
internal reference counting maintained by the library.
There are two reasons for this added level of abstraction.
First, it unifies the interface between scalar parameters \- parameters
that are constant across all frequencies (e.g. -1.0 for short) \- and
vector parameters that are given at a list of frequency points.
Second, it allows for parameters to be specified as unknown: parameters
that the library has to solve for.
.PP
All of the functions take a pointer to a \fBvnacal_t\fP obtained from
\fBvnacal_create\fP() or \fBvnacal_load\fP().  See \fBvnacal\fP(3).
.PP
\fBvnacal_make_scalar_parameter\fP() creates a frequency-independent
parameter with a reflection or transmission coefficient of
\fIcoefficient\fP.
.PP
\fBvnacal_make_vector_parameter\fP() creates a frequency-dependent
parameter, where \fIfrequency_vector\fP and \fIcoefficient_vector\fP are
vectors of length \fIfrequencies\fP giving the frequency and value
at each point.
The library uses rational function interpolation to interpolate between
frequencies if the given frequency points don't align with the calibration
frequencies.
.PP
\fBvnacal_make_calkit_parameter\fP() returns a parameter for a single-port
calibration kit standard.
.PP
\fBvnacal_make_calkit_parameter_matrix\fP() fills the caller-supplied
\fIparameter_matrix\fP with parameters for a one or two port calibration
kit standard and returns the number of ports of the standard.
For short, open and load standards, the resulting parameter matrix
has dimensions 1x1 and return value is 1; for the through standard,
the resulting parameter matrix has dimensions 2x2 and return value is 2.
The parameter matrix is indexed in standard row-major order.
The \fIparameter_matrix_size\fP parameter gives the size of
\fIparameter_matrix\fP in bytes and is used to protect against buffer
overrun.
Both calkit parameter functions take a pointer to the following structure
describing the standard:
.sp
.nf
.RS
\f[C]
typedef struct vnacal_calkit_data {
    vnacal_calkit_type_t vcd_type;        /* type of standard */
    uint32_t vcd_flags;                   /* flags */
    double vcd_offset_delay;              /* delay (s) */
    double vcd_offset_loss;               /* loss (\(*W/s) */
    double vcd_offset_z0;                 /* lossless Zc (\(*W) */
    double vcd_fmin;                      /* min frequency (Hz) */
    double vcd_fmax;                      /* max frequency (Hz) */

    /* then, depending on type, at most one of... */
    double complex vcd_l_coefficients[4]; /* short only */
    double complex vcd_c_coefficients[4]; /* open only */
    double complex vcd_zl;                /* load only (\(*W) */
} vnacal_calkit_data_t;
\f[]
.RE
.fi
.sp
\fBvcd_type\fP can be \s-2VNACAL_CALKIT_SHORT\s+2,
\s-2VNACAL_CALKIT_OPEN\s+2, \s-2VNACAL_CALKIT_LOAD\s+2, or
\s-2VNACAL_CALKIT_THROUGH\s+2.
\fBvcd_flags\fP is a bitwise-OR of flag values, only one of which
is currently defined.
The \s-2VNACAL_CKF_TRADITIONAL\s+2 flag instructs the library to use the
traditional transmission line model described in Agilent note AN-1287-11
where an approximation was used to avoid the need for complex square root.
By default, the library uses the Keysight revised transmission line model.
The three offset parameters are as described in Keysight documentation.
Note that the units are seconds, ohms per second, and ohms, respectively,
for consistency with the rest of the library.
The \fBvcd_fmin\fP and \fBvcd_fmax\fP parameters give the minimum and
maximum frequencies in Hz for which the standard is valid.
The remaining parameters are in a union (not shown), thus only one
can be valid at a time, depending on \fBvcd_type\fP.
For the short standard, \fBvcd_l_coefficients\fP are the coefficients
of a frequency polynomial modeling the inductance of the standard.
The elements of this array can also be accessed using \fBvcd_l0\fP,
\fBvcd_l1\fP. \fBvcd_l2\fP and \fBvcd_l3\fP.
Units of the coefficients are Henries, Henries/Hz, Henries/Hz^2 and
Henries/Hz^3, respectively.
For the open standard, \fBvcd_c_coefficients\fP are the coefficients
of a frequency polynomial modeling the capacitance of the standard.
The elements of this array can also be accessed using \fBvcd_c0\fP,
\fBvcd_c1\fP. \fBvcd_c2\fP and \fBvcd_c3\fP.
Units of the coefficients are Farads, Farads/Hz, Farads/Hz^2 and
Farads/Hz^3, respectively.
For the load standard, \fBvcd_zl\fP is the load impedance in ohms.
It can in general be complex, which can be valid over a small
frequency range.
.PP
\fBvnacal_make_data_parameter\fP() returns a parameter for a single-port
data standard.
The given \fBvnadata_t\fP object must have dimensions 1x1 and must
be in or convertible to scattering parameters.
The library implicitly renormalizes reference impedances of data
standards to those of the VNA ports.
Rational function interpolation is used to interpolate frequency points.
.PP
\fBvnacal_make_data_parameter_matrix\fP() fills the caller-supplied
\fIparameter_matrix\fP with parameters for an NxN data standard
and returns the number of ports of the standard.
The \fIparameter_matrix_size\fP parameter gives the size of
\fIparameter_matrix\fP in bytes and is used to protect against buffer
overrun.
.PP
\fBvnacal_load_data_parameter_matrix\fP() is the same as
\fBvnacal_make_data_parameter_matrix\fP() except that it takes,
\fIfilename\fP, the name of a Touchstone of NPD file to load
instead of a \fBvnadata_t\fP structure.
.PP
\fBvnacal_make_unknown_parameter\fP() creates a parameter with unknown
value, where \fIinitial_guess\fP is either one of the pre-defined
constants or any scalar, vector, calkit or data parameter (including
the matrix variants), giving a starting point for the solution.
.PP
\fBvnacal_make_correlated_parameter\fP() creates a parameter with unknown
value that is correlated with another (possibly unknown) parameter,
\fIother\fP, with per-frequency standard deviation of the difference
between them given in \fIsigma_vector\fP.
The \fIsigma_frequency_vector\fP argument gives the list of frequencies
where sigma values are specified, and \fIsigma_frequencies\fP gives the
length of both \fIsigma_vector\fP and \fIsigma_frequency_vector\fP.
This type of parameter is useful for stochastic modeling of connection
non-repeatability.
.PP
If \fIsigma_frequencies\fP is 1, then \fIsigma_frequency_vector\fP is
ignored, and \fIsigma_vector\fP[0] is taken as a frequency-independent
value.
Otherwise, the frequencies given in \fIsigma_frequencies\fP
must be positive increasing values that overlap with, but
don't necessarily have to be the same as, those given in either
\fBvnacal_new_set_frequency_vector\fP() or those given in a vector
parameter referred to by \fIother\fP.
The library uses natural cubic-spline interpolation to interpolate sigma
values between frequency points.
.PP
\fBvnacal_get_parameter_value\fP() returns the value of the parameter
at a given frequency.
If given a scalar parameter, \fBvnacal_get_parameter_value\fP() ignores
\fIfrequency\fP, and simply returns the fixed complex parameter value;
if given a vector parameter, it computes the value at the given frequency,
interpolating as necessary.
For vector parameters, the \fIfrequency\fP argument must lie within the
frequency range of the parameter.
This function cannot be used for calkit or data parameters, which require
explicit reference impedances; use \fBvnacal_eval_parameter\fP() or
\fBvnacal_eval_parameter_matrix\fP() instead.
If given an unknown or correlated parameter, this function returns the
most recent value computed by \fBvnacal_new_solve\fP(), or fails if the
parameter has not been solved.
.PP
\fBvnacal_eval_parameter\fP() is the same as \fBvnacal_get_parameter_value\fP()
except that it takes the reference impedance needed for calkit and data
parameters.
.PP
\fBvnacal_eval_parameter_matrix\fP() evaluates the parameter matrix
at the given frequency and stores the result in the user-supplied
\fIresult_matrix\fP.
If given a scalar parameter, \fBvnacal_eval_parameter_matrix\fP() ignores
\fIfrequency\fP, and simply returns the fixed complex parameter value;
if given a vector, calkit or data parameter, it computes the value at
the given frequency.
For vector, calkit and data parameters, \fIfrequency\fP must lie within
the valid frequency range of the parameter.
The \fIz0_vector\fP parameter gives the reference impedances for
each port, needed to evaluate calkit and data parameters.
.PP
\fBvnacal_parameter_matrix_to_data\fP() evaluates the parameter matrix
using the frequencies and reference impedances given in the \fBvnadata_t\fP
structure and stores the result back into the \fBvnadata_t\fP structure.
The \fIrows\fP and \fIcolumns\fP parameters give the dimensions of
\fIparameter_matrix\fP and must match the dimensions of the \fBvnadata_t\fP
object.
.PP
\fBvnacal_delete_parameter\fP() removes a parameter from the
\fBvnacal_t\fP structure.
Similarly, \fBvnacal_delete_parameter_matrix\fP() deletes all parameters
in the given \fIparameter_matrix\fP.
Entries are set to -1 as they're deleted.
Note that this function does not call \fBfree\fP(3) on \fIparameter_matrix\fP.
It's not an error to delete a parameter that has been added to a

debian/changelog
libvna (%%VERSION%%-1) unstable; urgency=medium

  * Initial Release

 -- Scott Guthridge <scott_guthridg@rompromity.net>  Sat, 26 Feb 2022 23:05:38 +0000


src/vnacal_parameter.c
/*
 * Vector Network Analyzer Library
 * Copyright © 2020-2026 D Scott Guthridge <scott_guthridge@rompromity.net>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "archdep.h"

#include <assert.h>
#include <complex.h>
#include <errno.h>
#include <math.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "vnacal_internal.h"


/*
 * _vnacal_format_sxx: write the name of an S parameter into buffer
 *   @cur: buffer of at least size SXX_BUFFER_ALLOC
 *   @end: one past the end of buffer
 *   @row: zero-based row
 *   @column: zero-based column
 */
char *_vnacal_format_sxx(char *cur, char *end, int row, int column)
{
    if (cur >= end) {
	return end;
    }
    (void)snprintf(cur, end - cur, "S%d%s%d",
	    row + 1,
	    row > 8 || column > 8 ? "_" : "",
	    column + 1);
    end[-1] = '\000';
    cur += strlen(cur);
    return cur;
}

/*
 * _vnacal_get_parameter_name: copy descriptive name for parameter into buffer
 *   @vpmrp: parameter struct
 *   @with_sxx: include "sxx of" in front of standard parameters
 *   @buffer: buffer of at least PARAMETER_BUFFER_ALLOC + 1 chars for result
 */
void _vnacal_get_parameter_name(const vnacal_parameter_t *vpmrp, bool with_sxx,
	char *buffer)
{
    char *cur = buffer;
    char *end = &buffer[PARAMETER_BUFFER_ALLOC];

    switch (vpmrp->vpmr_type) {
    case VNACAL_NEW:
    default:
	abort();

    case VNACAL_SCALAR:
	(void)snprintf(cur, end - cur, "scalar(%f",
		creal(vpmrp->vpmr_coefficient));
	end[-1] = '\000';
	cur += strlen(cur);
	if (cimag(vpmrp->vpmr_coefficient) != 0.0) {
	    (void)snprintf(cur, end - cur, "%+fj",
		    cimag(vpmrp->vpmr_coefficient));
	}
	end[-1] = '\000';
	cur += strlen(cur);
	(void)stpecpy(cur, end, ") parameter");
	return;

    case VNACAL_VECTOR:
	(void)stpecpy(cur, end, "vector parameter");
	return;

    case VNACAL_UNKNOWN:
	(void)stpecpy(cur, end, "unknown parameter");
	return;

    case VNACAL_CORRELATED:
	(void)stpecpy(cur, end, "correlated parameter");
	return;

    case VNACAL_CALKIT:
    case VNACAL_DATA:
    case VNACAL_DEEMBED:
    case VNACAL_EMBED:
	{
	    vnacal_standard_t *stdp = vpmrp->vpmr_stdp;

	    /*
	     * If the standard has more than one port and with_sxx was
	     * given, start with "Sxx of ", where xx describes the S
	     * parameter of the standard.
	     */
	    if (stdp->std_ports > 1 && with_sxx) {
		cur = _vnacal_format_sxx(cur, end,
			vpmrp->vpmr_row, vpmrp->vpmr_column);
		cur = stpecpy(cur, end, " of ");
	    }
	    (void)stpecpy(cur, end, stdp->std_name);
	}
	return;
    }
}

/*
 * _vnacal_get_parameter: return a pointer to the parameter
 *   @vcp: pointer returned from vnacal_create or vnacal_load
 *   @parameter: index of parameter
 *
 * Returns NULL if not found.  Caller must report the error.
 */
vnacal_parameter_t *_vnacal_get_parameter(const vnacal_t *vcp, int parameter)
{
    const vnacal_parameter_collection_t *vprmcp = &vcp->vc_parameter_collection;
    vnacal_parameter_t *vpmrp;

    if (parameter < 0 || parameter >= vprmcp->vprmc_allocation ||
	    (vpmrp = vprmcp->vprmc_vector[parameter]) == NULL ||
	    vpmrp->vpmr_deleted) {
	return NULL;
    }
    assert(vpmrp->vpmr_index == parameter);
    assert(vpmrp->vpmr_vcp == vcp);
    return vpmrp;
}

/*
 * _vnacal_get_parameter_frange: get frequency limits for the given parameter
 *   @vpmrp: pointer returned from _vnacal_get_parameter
 *   @fmin: address of double to receive minimum
 *   @fmax: address of double to receive maximum
 */
void _vnacal_get_parameter_frange(vnacal_parameter_t *vpmrp,
	double *fmin, double *fmax)
{
    vnacal_parameter_t *vpmrp_orig = vpmrp;

    for (;;) {
	switch (vpmrp->vpmr_type) {
	case VNACAL_SCALAR:
	    *fmin = 0.0;
	    *fmax = INFINITY;
	    break;

	case VNACAL_VECTOR:
	    *fmin = vpmrp->vpmr_frequency_vector[0];
	    *fmax = vpmrp->vpmr_frequency_vector[vpmrp->vpmr_frequencies - 1];
	    break;

	case VNACAL_UNKNOWN:
	case VNACAL_CORRELATED:
	    vpmrp = vpmrp->vpmr_other;
	    continue;

	case VNACAL_CALKIT:
	    {
		vnacal_standard_t *stdp = vpmrp->vpmr_stdp;
		vnacal_calkit_standard_t *cstdp =
		    (vnacal_calkit_standard_t *)stdp;

		*fmin = cstdp->cstd_calkit_data.vcd_fmin;
		*fmax = cstdp->cstd_calkit_data.vcd_fmax;
	    }
	    break;

	case VNACAL_DATA:
	    {
		vnacal_standard_t *stdp = vpmrp->vpmr_stdp;
		vnacal_data_standard_t *dstdp = (vnacal_data_standard_t *)stdp;

		*fmin = dstdp->dstd_frequency_vector[0];
		*fmax = dstdp->dstd_frequency_vector[
		    dstdp->dstd_frequencies - 1];
	    }
	    break;

	case VNACAL_EMBED:
	case VNACAL_DEEMBED:
	    {
		vnacal_standard_t *stdp = vpmrp->vpmr_stdp;
		vnacal_embed_standard_t *estdp =
		    (vnacal_embed_standard_t *)stdp;

		*fmin = estdp->estd_fmin;
		*fmax = estdp->estd_fmax;
	    }
	    break;

	default:
	    assert(!"unexpected parameter type");
	}
	break;
    }

    /*
     * If the original object is of type VNA_CORRELATED, then further
     * restrict the range based on the sigma frequencies.
     */
    if (vpmrp_orig->vpmr_type == VNACAL_CORRELATED &&
	    vpmrp_orig->vpmr_sigma_frequency_vector != NULL) {
	int sf = vpmrp_orig->vpmr_sigma_frequencies;
	double smin = vpmrp_orig->vpmr_sigma_frequency_vector[0];
	double smax = vpmrp_orig->vpmr_sigma_frequency_vector[sf - 1];

	if (smin > *fmin) {
	    *fmin = smin;
	}
	if (smax < *fmax) {
	    *fmax = smax;
	}
    }
}

/*
 * _vnacal_alloc_parameter: allocate a vnacal_parameter and return index
 *   @function: name of user-called function
 *   @vcp: pointer returned from vnacal_create or vnacal_load
 */
vnacal_parameter_t *_vnacal_alloc_parameter(const char *function, vnacal_t *vcp)
{
    vnacal_parameter_collection_t *vprmcp = &vcp->vc_parameter_collection;
    vnacal_parameter_t *vpmrp = NULL;
    int parameter;

    /*
     * Find a free slot in the table, extending the table if necessary.
     */
    if (vprmcp->vprmc_count < vprmcp->vprmc_allocation) {
	parameter = vprmcp->vprmc_first_free;
	while (vprmcp->vprmc_vector[parameter] != NULL) {
	    ++parameter;
	    assert(parameter < vprmcp->vprmc_allocation);
	}
	vprmcp->vprmc_first_free = parameter + 1;

    } else {
	vnacal_parameter_t **vpmrpp;
	int old_allocation = vprmcp->vprmc_allocation;
	int new_allocation;

	if (old_allocation < 3) {
	    new_allocation = 3;
	} else if (old_allocation < 8) {
	    new_allocation = 8;
	} else {
	    new_allocation = 2 * old_allocation;
	}
	if ((vpmrpp = realloc(vprmcp->vprmc_vector, new_allocation *
			sizeof(vnacal_parameter_t *))) == NULL) {
	    _vnacal_error(vcp, VNAERR_SYSTEM,
		    "realloc: %s", strerror(errno));
	    return NULL;
	}
	(void)memset((void *)&vpmrpp[old_allocation], 0,
		(new_allocation - old_allocation) *
		sizeof(vnacal_parameter_t *));
	vprmcp->vprmc_vector = vpmrpp;
	vprmcp->vprmc_allocation = new_allocation;
	parameter = vprmcp->vprmc_count;
    }

    /*
     * Allocate and init the new parameter.  Add it to the table.
     */
    if ((vpmrp = malloc(sizeof(vnacal_parameter_t))) == NULL) {
	_vnacal_error(vcp, VNAERR_SYSTEM,
		"malloc: %s", strerror(errno));
	return NULL;
    }
    (void)memset((void *)vpmrp, 0, sizeof(*vpmrp));
    vpmrp->vpmr_type = VNACAL_NEW;
    vpmrp->vpmr_deleted = false;
    vpmrp->vpmr_hold_count = 1;
    vpmrp->vpmr_index = parameter;
    vpmrp->vpmr_segment = 0;
    vpmrp->vpmr_vcp = vcp;
    vprmcp->vprmc_vector[parameter] = vpmrp;
    ++vprmcp->vprmc_count;
    return vpmrp;
}

/*
 * _vnacal_free_parameter: remove a parameter from the table and free
 *   @vpmrp: pointer returned from _vnacal_get_parameter
 */
static void _vnacal_free_parameter(vnacal_parameter_t *vpmrp)
{
    vnacal_t *vcp = vpmrp->vpmr_vcp;
    vnacal_parameter_collection_t *vprmcp = &vcp->vc_parameter_collection;
    int parameter = vpmrp->vpmr_index;

    assert(vpmrp != NULL);
    assert(vprmcp->vprmc_count >= 1);
    vprmcp->vprmc_vector[parameter] = NULL;
    --vprmcp->vprmc_count;
    if (parameter < vprmcp->vprmc_first_free) {
	vprmcp->vprmc_first_free = parameter;
    }
    switch (vpmrp->vpmr_type) {
    case VNACAL_NEW:
    case VNACAL_SCALAR:
	break;

    case VNACAL_CORRELATED:
	if (vpmrp->vpmr_sigma_frequency_vector !=
		vpmrp->vpmr_other->vpmr_frequency_vector) {
	    free((void *)vpmrp->vpmr_sigma_frequency_vector);
	}
	free((void *)vpmrp->vpmr_sigma_vector);
	free((void *)vpmrp->vpmr_sigma_spline);
	/*FALLTHROUGH*/

    case VNACAL_UNKNOWN:
	if (vpmrp->vpmr_other != NULL) {
	    _vnacal_release_parameter(vpmrp->vpmr_other);
	}
	/*FALLTHROUGH*/

    case VNACAL_VECTOR:
	free((void *)vpmrp->vpmr_frequency_vector);
	free((void *)vpmrp->vpmr_coefficient_vector);
	break;

    case VNACAL_CALKIT:
    case VNACAL_DATA:
    case VNACAL_EMBED:
    case VNACAL_DEEMBED:
	_vnacal_release_standard(&vpmrp->vpmr_stdp);
	break;

    default:
	abort();
    }
    free((void *)vpmrp);
}

/*
 * _vnacal_hold_parameter: increase the hold count on a parameter
 *   @vcp: pointer returned from vnacal_create or vnacal_load
 *   @parameter: index of parameter
 */
void _vnacal_hold_parameter(vnacal_parameter_t *vpmrp)
{
    ++vpmrp->vpmr_hold_count;
}

/*
 * _vnacal_release_parameter: decrease the hold count and free if zero
 *   @vcp: pointer returned from vnacal_create or vnacal_load
 *   @parameter: index of parameter
 */
void _vnacal_release_parameter(vnacal_parameter_t *vpmrp)
{
    assert(vpmrp->vpmr_hold_count > 0);
    if (--vpmrp->vpmr_hold_count == 0) {
	assert(vpmrp->vpmr_deleted);
	_vnacal_free_parameter(vpmrp);
    }
}

/*
 * _vnacal_setup_parameter_collection: allocate the parameter collection
 *   @function: name of user-called function
 *   @vcp: pointer returned from vnacal_create or vnacal_load
 */
int _vnacal_setup_parameter_collection(const char *function, vnacal_t *vcp)
{
    vnacal_parameter_t *vpmrp;

    /*
     * Zero the parameter collection structure.
     */
    (void)memset((void *)&vcp->vc_parameter_collection, 0,
	    sizeof(vcp->vc_parameter_collection));

    /*
     * Create the match parameter.
     */
    vpmrp = _vnacal_alloc_parameter(function, vcp);
    if (vpmrp == NULL) {
	goto error;
    }
    vpmrp->vpmr_type = VNACAL_SCALAR;
    vpmrp->vpmr_coefficient = 0.0;
    assert(vpmrp->vpmr_index == VNACAL_MATCH);

    /*
     * Create the open parameter.
     */
    vpmrp = _vnacal_alloc_parameter(function, vcp);
    if (vpmrp == NULL) {
	goto error;
    }
    vpmrp->vpmr_type = VNACAL_SCALAR;
    vpmrp->vpmr_coefficient = 1.0;
    assert(vpmrp->vpmr_index == VNACAL_OPEN);

    /*
     * Create the short parameter.
     */
    vpmrp = _vnacal_alloc_parameter(function, vcp);
    if (vpmrp == NULL) {
	goto error;
    }
    vpmrp->vpmr_type = VNACAL_SCALAR;
    vpmrp->vpmr_coefficient = -1.0;
    assert(vpmrp->vpmr_index == VNACAL_SHORT);

#if VNACAL_PREDEFINED_PARAMETERS != 3
#error "missing initializations in _vnacal_setup_parameter_collection"
#endif
    return 0;

error:
    _vnacal_teardown_parameter_collection(vcp);
    return -1;
}

#ifdef DEBUG
/*
 * print_parameter_collection: show the parameter collection for debug
 *   @vprmcp: parameter collection
 */
void print_parameter_collection(const vnacal_parameter_collection_t *vprmcp)
{
    char name[PARAMETER_BUFFER_ALLOC + 1];

    (void)printf("parameter collection:\n");
    for (int slot = 0; slot < vprmcp->vprmc_allocation; ++slot) {
	vnacal_parameter_t *vpmrp = vprmcp->vprmc_vector[slot];
	vnacal_standard_t *stdp;
	vnacal_embed_standard_t *estdp;

	(void)printf("  slot %d: ", slot);
	if (vpmrp == NULL) {
	    (void)printf("free\n");
	    continue;
	}
	_vnacal_get_parameter_name(vpmrp, /*with_sxx=*/true, name);
	(void)printf("%s\n", name);
	(void)printf("    hold_count %d", vpmrp->vpmr_hold_count);
	if (vpmrp->vpmr_deleted) {
	    (void)printf(" (DELETED)");
	}
	(void)printf("\n");
	assert(vpmrp->vpmr_index == slot);

	switch (vpmrp->vpmr_type) {
	case VNACAL_NEW:
	case VNACAL_VECTOR:
	case VNACAL_UNKNOWN:
	case VNACAL_CORRELATED:
	default:
	    continue;

	case VNACAL_SCALAR:
	    (void)printf("    value %f%+f\n",
		    creal(vpmrp->vpmr_coefficient),
		    cimag(vpmrp->vpmr_coefficient));
	    continue;

	case VNACAL_CALKIT:
	case VNACAL_DATA:
	    stdp = vpmrp->vpmr_stdp;
	    (void)printf("    standard @%p refcount %d\n",
		    (void *)stdp, stdp->std_refcount);
	    continue;

	case VNACAL_EMBED:
	case VNACAL_DEEMBED:
	    stdp = vpmrp->vpmr_stdp;
	    estdp = (vnacal_embed_standard_t *)stdp;
	    (void)printf("    standard @%p refcount %d\n",
		    (void *)stdp, stdp->std_refcount);
	    (void)printf("    target matrix:\n");
	    for (int row = 0; row < stdp->std_ports; ++row) {
		for (int column = 0; column < stdp->std_ports; ++column) {
		    const int cell = stdp->std_ports * row + column;
		    vnacal_parameter_t *vpmrp_temp;

		    (void)printf("        %d %d ", row, column);
		    vpmrp_temp = estdp->estd_target_matrix[cell];
		    if (vpmrp_temp == NULL) {
			(void)printf("null\n");
			continue;
		    }
		    _vnacal_get_parameter_name(vpmrp_temp,
			    /*with_sxx=*/true, name);
		    (void)printf("slot %d \"%s\"\n",
			    vpmrp_temp->vpmr_index, name);
		}
	    }
	    (void)printf("    fixture matrix:\n");
	    for (int row = 0; row < 2 * stdp->std_ports; ++row) {
		for (int column = 0; column < 2 * stdp->std_ports; ++column) {
		    const int cell = 2 * stdp->std_ports * row + column;
		    vnacal_parameter_t *vpmrp_temp;

		    (void)printf("        %d %d ", row, column);
		    vpmrp_temp = estdp->estd_fixture_matrix[cell];
		    if (vpmrp_temp == NULL) {
			(void)printf("null\n");
			continue;
		    }
		    _vnacal_get_parameter_name(vpmrp_temp,
			    /*with_sxx=*/true, name);
		    (void)printf("slot %d \"%s\"\n",
			    vpmrp_temp->vpmr_index, name);
		}
	    }
	    continue;
	}
    }
    (void)fflush(stdout);
}
#endif

/*
 * _vnacal_teardown_parameter_collection: free the parameter collection
 *   @vcp: pointer returned from vnacal_create or vnacal_load
 */
void _vnacal_teardown_parameter_collection(vnacal_t *vcp)
{
    vnacal_parameter_collection_t *vprmcp = &vcp->vc_parameter_collection;

#ifdef DEBUG
    print_parameter_collection(vprmcp);
#endif /* DEBUG */
    for (int i = vprmcp->vprmc_allocation - 1; i >= 0; --i) {
	vnacal_parameter_t *vpmrp = vprmcp->vprmc_vector[i];

	/*
	 * Delete any parameters the caller didn't delete.
	 */
	if (vpmrp != NULL && !vpmrp->vpmr_deleted) {
	    vpmrp->vpmr_deleted = true;
	    _vnacal_release_parameter(vpmrp);
	}
    }
#if DEBUG > 1
    print_parameter_collection(vprmcp);
#endif
    /* check for refcount/memory leaks */
    for (int i = vprmcp->vprmc_allocation - 1; i >= 0; --i) {
	assert(vprmcp->vprmc_vector[i] == NULL);
    }
    free((void *)vprmcp->vprmc_vector);
    (void)memset((void *)&vcp->vc_parameter_collection, 0,
	    sizeof(vcp->vc_parameter_collection));
}
